lite/inquiring_certifier.go

```diff
@@ -1,163 +1,216 @@
 	return &InquiringCertifier{
-		cert:    NewDynamicCertifier(chainID, fc.Validators, fc.Height()),
+		logger:  log.NewNopLogger(),
+		chainID: chainID,
 		trusted: trusted,
-		Source:  source,
-	}, nil
-}
-
-// ChainID returns the chain id.
-// Implements Certifier.
-func (ic *InquiringCertifier) ChainID() string {
-	return ic.cert.ChainID()
+		source:  source,
+	}
 }
 
-// Validators returns the validator set.
-func (ic *InquiringCertifier) Validators() *types.ValidatorSet {
-	return ic.cert.cert.vSet
+func (ic *InquiringCertifier) SetLogger(logger log.Logger) {
+	logger = logger.With("module", "lite")
+	ic.logger = logger
+	ic.trusted.SetLogger(logger)
+	ic.source.SetLogger(logger)
 }
 
-// LastHeight returns the last height.
-func (ic *InquiringCertifier) LastHeight() int64 {
-	return ic.cert.lastHeight
+// Implements Certifier.
+func (ic *InquiringCertifier) ChainID() string {
+	return ic.chainID
 }
 
-// Certify makes sure this is checkpoint is valid.
-//
-// If the validators have changed since the last know time, it looks
-// for a path to prove the new validators.
-//
-// On success, it will store the checkpoint in the store for later viewing
 // Implements Certifier.
-func (ic *InquiringCertifier) Certify(commit Commit) error {
-	err := ic.useClosestTrust(commit.Height())
+//
+// If the validators have changed since the last know time, it looks to
+// ic.trusted and ic.source to prove the new validators.  On success, it will
+// try to store the SignedHeader in ic.trusted if the next
+// validator can be sourced.
+func (ic *InquiringCertifier) Certify(shdr types.SignedHeader) error {
+
+	// Get the latest known full commit <= h-1 from our trusted providers.
+	// The full commit at h-1 contains the valset to sign for h.
+	h := shdr.Height - 1
+	tfc, err := ic.trusted.LatestFullCommit(ic.chainID, 1, h)
 	if err != nil {
 		return err
 	}
 
-	err = ic.cert.Certify(commit)
-	if !liteErr.IsValidatorsChangedErr(err) {
-		return err
-	}
-	err = ic.updateToHash(commit.Header.ValidatorsHash)
-	if err != nil {
-		return err
+	if tfc.Height() == h {
+		// Return error if valset doesn't match.
+		if !bytes.Equal(
+			tfc.NextValidators.Hash(),
+			shdr.Header.ValidatorsHash) {
+			return lerr.ErrUnexpectedValidators(
+				tfc.NextValidators.Hash(),
+				shdr.Header.ValidatorsHash)
+		}
+	} else {
+		// If valset doesn't match...
+		if !bytes.Equal(tfc.NextValidators.Hash(),
+			shdr.Header.ValidatorsHash) {
+			// ... update.
+			tfc, err = ic.updateToHeight(h)
+			if err != nil {
+				return err
+			}
+			// Return error if valset _still_ doesn't match.
+			if !bytes.Equal(tfc.NextValidators.Hash(),
+				shdr.Header.ValidatorsHash) {
+				return lerr.ErrUnexpectedValidators(
+					tfc.NextValidators.Hash(),
+					shdr.Header.ValidatorsHash)
+			}
+		}
 	}
 
-	err = ic.cert.Certify(commit)
+	// Certify the signed header using the matching valset.
+	cert := NewBaseCertifier(ic.chainID, tfc.Height()+1, tfc.NextValidators)
```

Comment:
Do we need a BaseCertifier data structure? Maybe it's sufficient
having Certify method which accepts as parameters chainId, height, valSet
and header to be checked. We are creating here structure just to be able to
check it and then we throw it away.

Reply:
I think it's fine for now.  We'll likely go through a few more refactors of this before it settles anyways, and for now it's more important to get it working.